### PR TITLE
Menu item organization

### LINKED ISSUE #2621: Menu - Add the ability to group items
When there are more than a couple of items in a menu and the items could be grouped to aid understanding them:
- how should actions be grouped?
- how should that grouping be indicated? separator, text heading?

## deliverables

- [x] [problem statement for this issue](https://github.com/grommet/hpe-design-system/issues/2621#issuecomment-1146373069)
- [x] [guidance on action grouping, group taxonomy, ordering of groups and items within groups, examples](https://github.com/grommet/hpe-design-system/issues/2621#issuecomment-1146378788)
- [x] grouping API for the Menu component (https://github.com/grommet/grommet/pull/6162)
- [x] [visual indication of groups](https://github.com/grommet/hpe-design-system/issues/2621#issuecomment-1146382004)
- [ ] Add to Design System site menu item grouping and ordering guidance and examples #2636

## possible grouping examples

- state changing actions: enable/disable, power on/off
- transformational actions: edit, access
- transfer actions: share, download
- dangerous actions: delete
- navigation

## possible Menu api

`Menu items={[[item1, item2], [item3, item4, item5], [item6]]} />`

## possible grouping indicators

- horizontal rule separating the items
- heading text

### PR DESCRIPTION
**Pull request by halocline**
[Deploy Preview](https://deploy-preview-2636--keen-mayer-a86c8b.netlify.app/components/menu#organizing-menu-items)

#### What does this PR do?

Closes #2621 by adding "Organizing menu items" section.

#### Where should the reviewer start?

/components/menu#organizing-menu-items

#### What testing has been done on this PR?

In addition to the feature you are implementing, have you checked the following:

**General UX Checks**
- [x] Small, medium, and large screen sizes
- [x] Cross-browsers (FireFox, Chrome, and Safari)
- [x] Light & dark modes
- [x] All hyperlinks route properly

**Accessibility Checks**
- [x] Keyboard interactions
- [x] Screen reader experience
- [x] Run WAVE accessibility plugin (Chrome)

**Code Quality Checks**
- [ ] Console is free of warnings and errors
- [ ] Passes E2E commit checks
- [x] Visual snapshots are reasonable

### CODE REVIEW

File: aries-site/src/pages/components/menu.mdx [hunk 1]
```diff
@@ -42,6 +47,164 @@ Accessibility features such as screen reader messages are automatically built in
...
+    {/* Box allows menu to be presented with drop open, overriding Example align/justify center */}
+    <Box height="xsmall">
+      <MenuGroupingExample />
+    </Box>
+  </Example>
+  <Example
+    bestPractice={{
+      type: 'dont',
+      message: `Don't add a separator between groups when there is only 1 menu item in 
+      each category. A single dangerous actions in the exception and should be placed
```
Comment (taysea) [inline aries-site/src/pages/components/menu.mdx:90]:
```suggestion
      each category. A single dangerous actions is the exception and should be placed 
```

File: aries-site/src/pages/components/menu.mdx [hunk 2]
```diff
@@ -42,6 +47,164 @@ Accessibility features such as screen reader messages are automatically built in
...
+#### Transferring
+
+Actions enabling a user to share or consume data related to a resource.
+
+- Download [resource] data
+- Share [resource] with [user]
+- Send [resource] to [user]
+- Notify [user] about [resource]
+
+TODO: add a series of examples showcasing tangible resources (e.g. servers, clusters, etc.)
```
Comment (taysea) [inline aries-site/src/pages/components/menu.mdx:185]: Delete this

File: aries-site/src/pages/components/menu.mdx [hunk 3]
```diff
@@ -42,6 +47,164 @@ Accessibility features such as screen reader messages are automatically built in
...
+
+#### Batch actions menu related to a selection of servers
```
Comment (taysea) [inline aries-site/src/pages/components/menu.mdx:187]: I liked the idea of adding a small description here about which about the above groupings this menu is representing / why one group is on top of the other

Comment (halocline) [reply aries-site/src/pages/components/menu.mdx:185]: just pushed changes which include descriptions

File: aries-site/src/pages/components/menu.mdx (hunk 2 above)
Comment (halocline) [reply aries-site/src/pages/components/menu.mdx]: just pushed changes with this included

File: aries-site/src/pages/components/menu.mdx [hunk 4]
```diff
@@ -42,6 +47,164 @@ Accessibility features such as screen reader messages are automatically built in
...
+  >
+    {/* Box allows menu to be presented with drop open, overriding Example align/justify center */}
+    <Box height="small">
+      <MenuDangerousExample />
+    </Box>
+  </Example>
+  <Example
+    bestPractice={{
+      type: 'dont',
+      message: `Don't leave dangerous actions grouped with other actions.`,
```
[the remark this answers is not in the document]
Comment (halocline) [reply aries-site/src/pages/components/menu.mdx]: done

File: aries-site/src/pages/components/menu.mdx [hunk 5]
```diff
@@ -42,6 +47,164 @@ Accessibility features such as screen reader messages are automatically built in
...
+</Example>
+
+#### Actions menu related to an individual record
+
+<Example
+  code="https://raw.githubusercontent.com/grommet/hpe-design-system/master/aries-site/src/examples/components/menu/MenuRecordActionsExample.js"
+  docs="https://v2.grommet.io/menu?theme=hpe#props"
+  height={{ min: 'medium' }}
+>
+  <MenuRecordActionsExample />
```
[the remark this answers is not in the document]
Comment (halocline) [reply aries-site/src/pages/components/menu.mdx:219]: removed icon and stacked # servers in column

**Review by taysea — APPROVED** [review]
LGTM! Think we can pass along to team for feedback.

**Comment by bahriaditi** [comment]
Hi @halocline - this is looking great, I think the documentation is pretty solid and easy to understand. Thanks for adding in the visual examples and "do's and don'ts)

I have only one comment regarding a visual design detail:
Now that we are adding a divider/ border in cases with item grouping, I'm feeling the need for a default border around the menu list container. This makes it feel more visually grounded somehow (soft opinion). I realize this might be scope creep so let me know if this warrants another ticket. Here's the difference it would make:

Screen Shot 2022-06-07 at 3 01 17 PM



@vavalos5 would love to hear your thoughts as well!

**Comment by ericsoderberghp** [comment]
> Hi @halocline - this is looking great, I think the documentation is pretty solid and easy to understand. Thanks for adding in the visual examples and "do's and don'ts)
> 
> I have only one comment regarding a visual design detail: Now that we are adding a divider/ border in cases with item grouping, I'm feeling the need for a default border around the menu list container. This makes it feel more visually grounded somehow (soft opinion). I realize this might be scope creep so let me know if this warrants another ticket. Here's the difference it would make:
> 
> Screen Shot 2022-06-07 at 3 01 17 PM
> 
> @vavalos5 would love to hear your thoughts as well!

Would we then want consistency with other drop downs? For instance, Select options, TextInput suggestions, etc.?

**Comment by vavalos5** [comment]
> Hi @halocline - this is looking great, I think the documentation is pretty solid and easy to understand. Thanks for adding in the visual examples and "do's and don'ts)
> 
> I have only one comment regarding a visual design detail: Now that we are adding a divider/ border in cases with item grouping, I'm feeling the need for a default border around the menu list container. This makes it feel more visually grounded somehow (soft opinion). I realize this might be scope creep so let me know if this warrants another ticket. Here's the difference it would make:
> 
> Screen Shot 2022-06-07 at 3 01 17 PM
> 
> @vavalos5 would love to hear your thoughts as well!

I would agree with Aditi on this looking awesome - thanks for doing this Matt! It is easy to understand and I like the do's and don'ts examples. This is probably a separate ticket but I had a hard time understanding when I should use a border around the menu button vs. not. I know the examples with the borders are showcasing our toolbar button which should only be used at the toolbar area ( I know we have this guidance in button but we should probably link that somewhere in here). Lmk if I should make a ticket for that.

In regards to adding a border as @bahriaditi suggested. I like it but at the same time, i'm wondering if we'd then want to apply that to many of our other components like modals, layers, PageHeader, etc.. I think i'd have to let is sit for another day before coming to conclusion just because many of our components don't have an outline and i'm afraid this will start creating more visual noise? Just thinking out loud.

**Comment by ericsoderberghp** [comment]
I'm also wondering about the individual Server Groups having borders around them rather than using Card or List and relying on the existing theme styling for them.

**Comment by bahriaditi** [comment]
@ericsoderberghp + @vavalos5 agreed that we would then need visual consistency and apply borders to other similar components. Let's hold off on this - I can create a ticket to revisit later. Definitely not a must-have / blocker for this work.

**Comment by halocline** [comment]
> I have only one comment regarding a visual design detail: Now that we are adding a divider/ border in cases with item grouping, I'm feeling the need for a default border around the menu list container. This makes it feel more visually grounded somehow (soft opinion). I realize this might be scope creep so let me know if this warrants another ticket. Here's the difference it would make:
> 

We could also explore a divider which doesn't span the entire width of the drop container. That way the divider doesn't touch the edges and may alleviate the urge for a border.

**Comment by halocline** [comment]
> I'm also wondering about the individual Server Groups having borders around them rather than using Card or List and relying on the existing theme styling for them.

Removed.

Screen Shot 2022-06-08 at 8 33 10 AM

**Comment by vavalos5** [comment]
> > This is probably a separate ticket but I had a hard time understanding when I should use a border around the menu button vs. not. I know the examples with the borders are showcasing our toolbar button which should only be used at the toolbar area ( I know we have this guidance in button but we should probably link that somewhere in here). Lmk if I should make a ticket for that.
> 
> Guidance for when to use toolbar menu exists here: https://http://design-system.hpe.design/components/menu#within-a-toolbar
> 
> Does this address what you are wanting?

Hey Matt, that link doesn't seem to work. But if we're just referring to the toolbar section within Button then yes that's what i'm thinking we should address within Menu. Just when to use one or the other. Or linking int

**Comment by halocline** [comment]
> Hey Matt, that link doesn't seem to work. But if we're just referring to the toolbar section within Button then yes that's what i'm thinking we should address within Menu. Just when to use one or the other. Or linking int

Link fixed. This guidance exists on the Menu page.  https://design-system.hpe.design/components/menu#within-a-toolbar

File: aries-site/src/pages/components/menu.mdx [hunk 6]
```diff
@@ -42,6 +47,178 @@ Accessibility features such as screen reader messages are automatically built in
...
+
+<Example
+  code={[
+    "https://raw.githubusercontent.com/grommet/hpe-design-system/master/aries-site/src/examples/components/menu/MenuRecordActionsExample.js",
+    "https://raw.githubusercontent.com/grommet/hpe-design-system/master/aries-site/src/data/mockData/serverGroups.json"
+  ]}
+  github="https://github.com/grommet/hpe-design-system/blob/master/aries-site/src/examples/components/menu/MenuRecordActionsExample.js"
+  height={{ min: 'medium' }}
+>
+  <MenuRecordActionsExample />
```
[the remark this answers is not in the document]
Comment (halocline) [reply aries-site/src/pages/components/menu.mdx:219]: There's a lot of overhead to implement the double confirmation, but yes.

I made a reusable template called DestructiveConfirmation applied here, the destructive action example, and to where it is used on the Content Layouts page.

File: aries-site/src/pages/components/menu.mdx [hunk 7]
```diff
@@ -42,6 +47,178 @@ Accessibility features such as screen reader messages are automatically built in
...
+Actions enabling a user to share or consume data related to a resource.
+
+- Download [resource] data
+- Share [resource] with [user]
+- Send [resource] to [user]
+- Notify [user] about [resource]
+
+#### Batch actions menu related to a selection of servers
+
+The table menu's batch actions are grouped by exectuing and configuration actions.
```
[the remark this answers is not in the document]
Comment (halocline) [reply aries-site/src/pages/components/menu.mdx]: done

**Comment by dahliaferrer** [comment]
This sentence confused me until I read the rest of the guidance... 15 minutes later LOL

"Create separate groups for menu items initiating actions and items setting attributes. Menu items become unpredictable when the behavior of grouped items varies."

Suggest:
Create separate groups for menu items **that initiate actions and set attributes**...

Tiny typo:
A single dangerous action**s** is the exception and should be placed in a distinct group.
[remove the "s"]

Delete profile interactive action in the Dangerous section:
I have a screenshot but I can't figure out how to attach it here. Anyway, when you interact with this and click on "delete" we seem to give guidance that the user must enter this long string to delete. But in our Forms guide, we have an interactive example to simply write the word "DELETE," and then click on a delete button below. If that "delete" guidance is correct, then could we change out this one so that it matches our guidance?

**Pull request merged**